File: Code/Ch1-memory-centric-system-model/cpp-code/abstract_system_model.cc
#include <iostream>
#include <tuple>

#define True 1
#define False 0
/*
A model for a processor-based system. The main difference from the code in the book is that it uses NumPy arrays for performance.;
Furthermore,the model for loading high;
*/

#include "abstract_system_constants.h"
#include "abstract_system_peripherals.h"
#include "abstract_system_processor.h"
#include "abstract_system_code.h"

//#include "sys.h"

int main() {

    // Default settings
    const uint64_t     PRINT_CYCLES=False;
    const uint64_t     MAX_NCYCLES=512;
    // TODO: cmd line args

    // Initialise the state

    //systemState=[0]*systemStateSz
    std::vector<uint64_t> systemState;//(systemStateSz, 0);

    std::vector<uint64_t> timerState(timerStateSz,0);
    timerState[0]=MAX_NCYCLES+1;
    timerState[1]=MAX_NCYCLES+1;
    std::vector<uint64_t> kbdState(kbdStateSz,0);
    std::vector<uint64_t> nicState(nicStateSz,0);
    std::vector<uint64_t> ssdState(ssdStateSz,0);
    std::vector<uint64_t> gpuState(gpuStateSz,0);
    std::vector<uint64_t> ramState(ramStateSz);
    std::vector<uint64_t> irqs;
    uint64_t kbdIrq=0;
    uint64_t nicIrq=0;
    uint64_t ssdIrq=0;
    uint64_t gpuIrq=0;
    uint64_t timerIrq=0;    

    // Initialise the registers
    std::vector<uint64_t> registers(NREGS,0);//#[0]*16;
    // Set the program counter to the start of the code region
    registers[PC]=CODE;
    std::vector<uint64_t> ivt(IVTsz,CODE);

    // Initialise the intterupt vector table
    for (int idx = IVT; idx<IVT+IVTsz;idx++) {
        ramState[idx]=CODE;
    }

    ramState=load_code(ramState);

    int ct=0;
    for (auto elt : ramState ) {
        systemState.push_back(elt);
    }
    for (auto elt : timerState ) {
        systemState.push_back(elt);
    }
    for (auto elt : kbdState ) {
        systemState.push_back(elt);
    }
    for (auto elt : nicState ) {
        systemState.push_back(elt);
    }
    for (auto elt : ssdState ) {
        systemState.push_back(elt);
    }
    for (auto elt : gpuState ) {
        systemState.push_back(elt);
    }

    for (int ncycles = 0; ncycles<MAX_NCYCLES;ncycles++ ) {
        if (PRINT_CYCLES ) {
            std::cout << ncycles << "\n";
        }
#ifdef PERIPHERALS
        std::tie(kbdState,kbdIrq)=kbdAction(kbdState);
        std::tie(nicState,nicIrq)=nicAction(nicState);
        std::tie(ssdState,ssdIrq)=ssdAction(ssdState);
        std::tie(gpuState,gpuIrq)=gpuAction(gpuState);
        std::tie(timerState,timerIrq)=timerAction(timerState);
        std::vector<uint64_t>::const_iterator first = systemState.begin() ;
        std::vector<uint64_t>::const_iterator last = systemState.begin() + MEMTOP;
        std::vector<uint64_t> ramState(first,last);//=systemState[0:MEMTOP];

        irqs={kbdIrq,nicIrq,ssdIrq,gpuIrq,timerIrq};

        //                systemState=ramState+timerState+kbdState+nicState+ssdState+gpuState
        //systemState=concatenate((ramState,timerState,kbdState,nicState,ssdState,gpuState)); // FIXME
#else 
        irqs={};
        //            print(systemState[CODE:CODE+20])
#endif
        std::tie(systemState,irqs,registers)=processorAction(systemState,irqs,registers);

    }
    std::cout << IVT+IVTsz <<" .. " <<IVT+IVTsz+FIB_MAX<<"\n";
    for (int idx = IVT+IVTsz; idx<IVT+IVTsz+FIB_MAX;idx++) {
        std::cout <<idx<<"\t"<<systemState[idx]<<std::endl;
    }
//   std::cout <<systemState[IVT+IVTsz:FIB_MAX])<<std::endl;

}

File: Code/Ch1-memory-centric-system-model/advanced-model/abstract_system_model.py
#!/usr/bin/env python
"""
A model for a processor-based system. 
The main difference from the code in the book is that it uses NumPy arrays for performance.
The model for loading high-level code is different as well.
"""


from abstract_system_constants import *
from abstract_system_peripherals import *
from abstract_system_processor import processorAction
from abstract_system_code import * #load_code

import numpy as np
import sys, getopt

def main():
    global PERIPHERALS, HL
    global hlCode

# Default settings
    PERIPHERALS=False
    HL=False
    PRINT_CYCLES=False

    MAX_NCYCLES= 512
    if HL:
        MAX_NCYCLES= 1

# Command line flags and arguments
    try:
        opts, args = getopt.getopt(sys.argv[1:],"hPHsN:")
    except getopt.GetoptError:
        print( 'abstract_system_model.py. [-hPHN:s]')
        sys.exit(2)

    for opt, arg in opts:
      if opt == '-h':
         print( 'abstract_system_model.py [-hPHc] -N ncycles'+"\n", 
            '-P : with peripherals'+"\n", 
            '-H : high-level (not cycle-accurate)'+"\n",
            '-N : number of cycles to run'+"\n",
            '-s: show cycle count'
            )
         sys.exit()
      elif opt in ("-P"):
         PERIPHERALS=True
      elif opt in ("-H"):
         HL = True
      elif opt in ('-s'):
          PRINT_CYCLES=True
      elif opt in ('-N'):
          MAX_NCYCLES=int(arg)

# If we use high-level code we need to store it in a separate array 
# because the NumPy arrays are declared for uint64
    # hlCode=[]
    # if HL:
        # Fill the code array with no-op functions
    hlCode=[lambda s,r: (s,r) ]*hlCodeSz

    # Configuration of the model
    cfg=(PERIPHERALS,HL,hlCode)

# Initialise the state

    #systemState= [0]*systemStateSz # without numpy
    systemState=np.zeros( systemStateSz,dtype=np.uint64 ) 

    timerState=np.zeros( timerStateSz,dtype=np.uint64 )
    timerState[0]=MAX_NCYCLES+1
    timerState[1]=MAX_NCYCLES+1
    kbdState=np.zeros( kbdStateSz,dtype=np.uint64 )
    nicState=np.zeros( nicStateSz,dtype=np.uint64 )
    ssdState=np.zeros( ssdStateSz,dtype=np.uint64 )
    gpuState=np.zeros( gpuStateSz,dtype=np.uint64 )
    dmaState=np.zeros( dmaStateSz,dtype=np.uint64 )
    ramState=np.zeros( ramStateSz,dtype=np.uint64 )

# Initialise the registers
    registers=np.zeros( NREGS,dtype=np.uint64 ) #[0]*NREGS
# Set the program counter to the start of the code region    
    registers[PC]=CODE

# Initialise the intterupt vector table    
    ramState[IVT:IVT+IVTsz]=[CODE+18]*IVTsz # TODO: use IRS addresses

    (ramState,cfg) = load_code(ramState,cfg)

    #systemState = ramState+timerState+kbdState+nicState+ssdState+gpuState
    systemState = np.concatenate((ramState,timerState,kbdState,nicState,ssdState,gpuState,dmaState))
    for ncycles in range(0,MAX_NCYCLES):
            if PRINT_CYCLES:
                print(ncycles)
            if PERIPHERALS:
                (kbdState,kbdIrq)=kbdAction(kbdState)
                (nicState,nicIrq)=nicAction(nicState)
                (ssdState,ssdIrq)=ssdAction(ssdState)
                (gpuState,gpuIrq)=gpuAction(gpuState)
                (timerState,timerIrq)=timerAction(timerState)
                (systemState,dmaIrq) = dmaAction(systemState)
                
                ramState=systemState[0:MEMTOP]

                irqs=[kbdIrq,nicIrq,ssdIrq,gpuIrq,timerIrq,dmaIrq]
                
#                systemState = ramState+timerState+kbdState+nicState+ssdState+gpuState # without numpy
                systemState = np.concatenate((ramState,timerState,kbdState,nicState,ssdState,gpuState,dmaState))
                timerState = systemState[TIMER:TIMER+timerStateSz-1]
                # TODO: other peripherals
            else:
                irqs=[]
#            print( systemState[CODE:CODE+20] )
            (systemState,irqs,registers) = processorAction(systemState,irqs,registers,cfg)

    print(systemState[IVT+IVTsz:FIB_MAX])

############ ############ ############ ############ ############ ############ ############ ############ 

main()
